### PR TITLE
JS: permit some calls with spurious arguments to empty functions

### PR DESCRIPTION
**Pull request by esben-semmle**
This fixes ODASA-7272 by not flagging calls to some functions without bodies, they are presumably debug functions with commented out bodies.

### CODE REVIEW

File: javascript/ql/src/LanguageFeatures/SpuriousArguments.ql [hunk 1]
```diff
@@ -83,5 +83,13 @@ class SpuriousArguments extends Expr {
 
 from SpuriousArguments args, Function f, string arguments
 where f = args.getCall().getACallee() and
-      if args.getCount() = 1 then arguments = "argument" else arguments = "arguments"
+      if args.getCount() = 1 then arguments = "argument" else arguments = "arguments" and
+      (
+        // exclude empty functions, they are probably commented out debug utilities ...
+        exists(f.getABodyStmt()) or
+        // ... but include, constructors, arrows and externals
+        any(ConstructorDeclaration cd).getBody() = f or
```
Comment (xiemaisi) [inline javascript/ql/src/LanguageFeatures/SpuriousArguments.ql]: `f instanceof Constructor`

File: javascript/ql/src/LanguageFeatures/SpuriousArguments.ql [hunk 2]
```diff
@@ -83,5 +83,13 @@ class SpuriousArguments extends Expr {
 
 from SpuriousArguments args, Function f, string arguments
 where f = args.getCall().getACallee() and
-      if args.getCount() = 1 then arguments = "argument" else arguments = "arguments"
+      if args.getCount() = 1 then arguments = "argument" else arguments = "arguments" and
+      (
+        // exclude empty functions, they are probably commented out debug utilities ...
+        exists(f.getABodyStmt()) or
+        // ... but include, constructors, arrows and externals
```
Comment (xiemaisi) [inline javascript/ql/src/LanguageFeatures/SpuriousArguments.ql]: Why are arrows special? And what about ambient TypeScript functions?

Comment (esben-semmle) [reply javascript/ql/src/LanguageFeatures/SpuriousArguments.ql]: > Why are arrows special? 

Arrows can not be empty, but they do not have to contain statements. 
Have you seen "practically empty" arrows in the wild that you want a heuristic for?

**Comment by esben-semmle** [comment]
Comments addressed with a squash.

File: javascript/ql/src/LanguageFeatures/SpuriousArguments.ql [hunk 3]
```diff
@@ -83,5 +83,14 @@ class SpuriousArguments extends Expr {
...
 from SpuriousArguments args, Function f, string arguments
 where f = args.getCall().getACallee() and
-      if args.getCount() = 1 then arguments = "argument" else arguments = "arguments"
+      if args.getCount() = 1 then arguments = "argument" else arguments = "arguments" and
+      (
+        // exclude empty functions, they are probably commented out debug utilities ...
+        exists(f.getABodyStmt()) or
+        // ... but include: constructors, arrows and externals/ambients
+        f instanceof Constructor or // unlikely to be a debug utility
+        f instanceof ArrowFunctionExpr or // can not be empty
```
Comment (xiemaisi) [inline javascript/ql/src/LanguageFeatures/SpuriousArguments.ql]: "cannot", please

File: javascript/ql/src/ql [hunk 4]
```diff
@@ -0,0 +1 @@
+/home/esben/_data/uber-poc/poc3/ql/
```
Comment (xiemaisi) [inline javascript/ql/src/ql]: ?

**Comment by esben-semmle** [comment]
Comments addressed with one more squash.

**Pull request merged**